bn-apps/ledger-sync/ledger-sync-service/src/test/kotlin/com/r3/businessnetworks/ledgersync/LedgerConsistencyTests.kt

```diff
@@ -189,6 +193,96 @@ class LedgerConsistencyTests {
         assertEquals(3, node1.fromNetwork().bogusStateCount())
     }
 
+    @Test
+    fun `only one node has been recycled` () {
+        node1.fromNetwork().createTransactions(10)
+
+        assertEquals(20, node1.fromNetwork().bogusStateCount())
+
+        // part of the transactions have been recycled
+        node1.fromNetwork().simulateVR(4)
+
+        assertEquals(16, node1.fromNetwork().bogusStateCount())
+
+        val ledgerSyncResult = node1.fromNetwork().runRequestLedgerSyncFlow(node1.fromNetwork().regularNodes())
+
+        assertEquals(0, ledgerSyncResult[node2.fromNetwork().identity()]!!.missingAtRequester.size)
+        assertEquals(0, ledgerSyncResult[node3.fromNetwork().identity()]!!.missingAtRequester.size)
+
+        // part of the transactions have been recycled, and the rest are lost
+        node1.fromNetwork().simulateCatastrophicFailure()
+
+        assertEquals(0, node1.fromNetwork().bogusStateCount())
+
+        val ledgerSyncResult2 = node1.fromNetwork().runRequestLedgerSyncFlow(node1.fromNetwork().regularNodes())
+
+        assertEquals(16, ledgerSyncResult2[node2.fromNetwork().identity()]!!.missingAtRequester.size + ledgerSyncResult2[node3.fromNetwork().identity()]!!.missingAtRequester.size)
+
+        // recovery
+        node1.fromNetwork().runTransactionRecoveryFlow(ledgerSyncResult2)
+
+        assertEquals(16, node1.fromNetwork().bogusStateCount())
+    }
+
+    @Test
+    fun `both nodes have been recycled on different transactions` () {
```

Comment:
multiple tests like this one. Notice that native SQL has also been used in test like line 402 and line 418.